ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/ExpressionTreeRewriter.java

```diff
@@ -465,6 +469,30 @@ public Expression visitCast(final Cast node, final C context) {
       return new Cast(node.getLocation(), expression, type);
     }
 
+    @Override
+    public Expression visitLambdaExpression(final LambdaFunctionExpression node, final C context) {
+      final Optional<Expression> result
+          = plugin.apply(node, new Context<>(context, this));
+      if (result.isPresent()) {
+        return result.get();
+      }
+
+      final LambdaContext lambdaContext =
```

Comment:
`LambdaContext` doesn't belong here - the rewriter shouldn't make any interpretation of the context. From this class's POV context is an opaque cookie to pass around to the rewriter so it can keep some internal state.

What's the intent of this LambdaContext stuff? is it just to screen for conflicts in variable names w/ enclosing variable names or column names? The better way to structure this would be to keep track of these in `AstSanitizer`. So from there you could do:

```
class AstSanitizer {
...
    private static final class ExpressionRewriterPlugin extends
      VisitParentExpressionVisitor<Optional<Expression>, Context<SanitizerContext>> {

          @Override
          public Optional<Expression> visitUnqualifiedColumnReference(
              final UnqualifiedColumnReferenceExp expression,
              final SanitizerContext ctx
          ) {
              final ColumnName columnName = expression.getColumnName();
              if (ctx.lambdaArgs().contains(columnName.text())) {
                  return Optional.of(new LambdaLiteral(columnName.text()));
              }
              ...
          }

          @Override
          public Optional<Expression> visitLambdaExpression(
              final LambdaFunctionExpression expression,
              final SanitizerContext ctx
           ) {
              dataSourceExtractor.getAllSources().forEach(aliasedDataSource -> {
              for (String argument : expression.getArguments()) {
                  if (aliasedDataSource.getDataSource().getSchema().columns().stream()
                      .map(column -> column.name().text()).collect(Collectors.toList())
                      .contains(argument)) {
                       throw new KsqlException("Lambda argument can't be a column name.");
                  }
                  if (ctx.lambdaArgs().contains(argument)) {
                       throw new KsqlException...
                  }
                  ctx.addLambdaArg(argument);
              }
              });
              return Optional.empty();  // just let the rewriter reconstruct the lambda - this function just validates and tracks args
          }
    }

    private static class SanitizerContext {
        Set<String> lambdaArgs = new HashSet<>();

        private void addLambdaArg(final String name) {
            reservedNames.add(name);
        }

        private Set<String> lambdaArgs() {
            return ImmutableSet.copyOf(lambdaArgs);
        }
    }
}
```

Then, this method should just become a dumb rewrite like everything else, and we can get rid of `LambdaContext`

Reply:
Sounds good to me. Then in https://github.com/confluentinc/ksql/pull/6868/files#diff-e5f9abbf3a2878a1dce09039136805764ecc31404a7eb27f2d1b49b9e038030dR81 we would just pass in a `new SanitizerContext()` instead of a `null`.

Reply:
I suppose we could, though tracking the lambda args seems specific to `AstSanitizer` at the moment.